src/python/pants/backend/python/macros/python_requirements.py

```diff
@@ -43,6 +45,40 @@
 from pants.util.strutil import softwrap
 
 
+def parse_pyproject_toml(
+    pyproject_toml: str,
+    *,
+    rel_path: str,
+    overrides: MutableMapping[NormalizedName, dict[str, Any]],
+) -> Iterator[PipRequirement]:
+    parsed = toml.loads(pyproject_toml)
+    deps_vals: list[str] = parsed.get("project", {}).get("dependencies", [])
+    optional_dependencies = parsed.get("project", {}).get("optional-dependencies", {})
+    if not deps_vals and not optional_dependencies:
+        raise KeyError(
+            softwrap(
+                "No section `project.dependencies` or `project.optional-dependencies` "
+                f"found in {rel_path}"
+            )
+        )
+    for dep in deps_vals:
+        dep, _, _ = dep.partition("--")
+        dep = dep.strip().rstrip("\\")
+        if not dep or dep.startswith(("#", "-")):
+            continue
+        yield PipRequirement.parse(dep, description_of_origin=rel_path)
+    for tag, opt_dep in optional_dependencies.items():
+        for dep in opt_dep:
+            req = PipRequirement.parse(dep, description_of_origin=rel_path)
+            canonical_project_name = canonicalize_project_name(req.project_name)
+            override = overrides.get(canonical_project_name, {})
+            tags: list[str] = override.get("tags", [])
+            tags.append(tag)
+            override["tags"] = tags
+            overrides[canonical_project_name] = override
+            yield req
```

Comment:
This code just adds the key from the `optional-dependencies` table as a tag to that requirement. I'm not sure if this is a good way to go about it, because in resolving #12487 it'll need to have some way to put things back into the right places in a generated `pyproject.toml`. But this is maybe good enough for now, just so this information is not totally thrown away.